cdap-proto/src/main/java/co/cask/cdap/proto/Id.java

```diff
@@ -379,6 +380,9 @@ public int hashCode() {
     private final String streamName;
     private transient int hashCode;
 
+    private transient String id;
```

Comment:
namespace and streamName are the core information of the class. the `id` field is more of a cached string computed from those two feilds. This way, whenever storing it, we just store only enough information as necessary.